File: clients/javascript/sync/tests/test_functional.mjs
// This is free software for the public good of a permacomputer hosted at
// permacomputer.com, an always-on computer by the people, for the people.
// One which is durable, easy to repair, & distributed like tap water
// for machine learning intelligence.
//
// The permacomputer is community-owned infrastructure optimized around
// four values:
//
//   TRUTH      First principles, math & science, open source code freely distributed
//   FREEDOM    Voluntary partnerships, freedom from tyranny & corporate control
//   HARMONY    Minimal waste, self-renewing systems with diverse thriving connections
//   LOVE       Be yourself without hurting others, cooperation through natural law
//
// This software contributes to that vision by enabling code execution across 42+ programming languages through a unified interface, accessible to all.
// Code is seeds to sprout on any abandoned technology.

/**
 * UN JavaScript SDK - Functional Tests
 *
 * Tests library functions against real API.
 * Requires: UNSANDBOX_PUBLIC_KEY, UNSANDBOX_SECRET_KEY
 *
 * Usage:
 *   node clients/javascript/sync/tests/test_functional.mjs
 */

import {
  executeCode,
  getLanguages,
  listSessions,
  createSession,
  deleteSession,
  listServices,
  listSnapshots,
  listImages,
  validateKeys,
  healthCheck,
} from '../src/un.js';

const GREEN = '\x1b[32m';
const RED = '\x1b[31m';
const BLUE = '\x1b[34m';
const YELLOW = '\x1b[33m';
const NC = '\x1b[0m';

let passed = 0;
let failed = 0;

function check(condition, msg) {
  if (condition) {
    console.log(`  ${GREEN}✓${NC} ${msg}`);
    passed++;
  } else {
    console.log(`  ${RED}✗${NC} ${msg}`);
    failed++;
  }
}

async function testHealthCheck() {
  console.log('\nTesting healthCheck()...');
  const result = await healthCheck();
  check(typeof result === 'boolean', 'healthCheck returns boolean');
}

async function testValidateKeys() {
  console.log('\nTesting validateKeys()...');
  const info = await validateKeys();
  check(info != null, 'validateKeys returns non-null');
  check(info.valid === true, 'keys are valid');
  if (info.tier) console.log(`    tier: ${info.tier}`);
}

async function testGetLanguages() {
  console.log('\nTesting getLanguages()...');
  const langs = await getLanguages();
  check(Array.isArray(langs), 'getLanguages returns array');
  check(langs.length > 0, 'at least one language returned');
  check(langs.includes('python'), 'python is in languages list');
  console.log(`    Found ${langs.length} languages`);
}

async function testExecute() {
  console.log('\nTesting executeCode()...');
  const result = await executeCode('python', "print('hello from JS SDK')");
  check(result != null, 'execute returns non-null');
  check(result.stdout && result.stdout.includes('hello from JS SDK'), 'stdout contains expected output');
  check(result.exit_code === 0, 'exit code is 0');
}

async function testExecuteError() {
  console.log('\nTesting executeCode() with error...');
  const result = await executeCode('python', 'import sys; sys.exit(1)');
  check(result != null, 'execute returns non-null');
  check(result.exit_code === 1, 'exit code is 1');
}

async function testSessionList() {
  console.log('\nTesting listSessions()...');
  const sessions = await listSessions();
  check(Array.isArray(sessions), 'listSessions returns array');
  console.log(`    Found ${sessions.length} sessions`);
}

async function testSessionLifecycle() {
  console.log('\nTesting session lifecycle (create, destroy)...');

  const session = await createSession('python');
  check(session != null, 'createSession returns non-null');
  const sessionId = session.session_id || session.id;
  check(sessionId != null, 'session has id');
  console.log(`    session_id: ${sessionId}`);

  if (sessionId) {
    const destroyed = await deleteSession(sessionId);
    check(destroyed != null, 'deleteSession returns non-null');
  }
}

async function testServiceList() {
  console.log('\nTesting listServices()...');
  const services = await listServices();
  check(Array.isArray(services), 'listServices returns array');
  console.log(`    Found ${services.length} services`);
}

async function testSnapshotList() {
  console.log('\nTesting listSnapshots()...');
  const snapshots = await listSnapshots();
  check(Array.isArray(snapshots), 'listSnapshots returns array');
  console.log(`    Found ${snapshots.length} snapshots`);
}

async function testImageList() {
  console.log('\nTesting listImages()...');
  const images = await listImages();
  check(Array.isArray(images), 'listImages returns array');
  console.log(`    Found ${images.length} images`);
}

// Main
async function main() {
  console.log('=====================================');
  console.log('UN JavaScript SDK - Functional Tests');
  console.log('Testing against real API');
  console.log('=====================================');

  if (!process.env.UNSANDBOX_PUBLIC_KEY || !process.env.UNSANDBOX_SECRET_KEY) {
    console.log(`\n${YELLOW}SKIP: UNSANDBOX_PUBLIC_KEY and UNSANDBOX_SECRET_KEY not set${NC}`);
    process.exit(0);
  }

  const tests = [
    testHealthCheck,
    testValidateKeys,
    testGetLanguages,
    testExecute,
    testExecuteError,
    testSessionList,
    testSessionLifecycle,
    testServiceList,
    testSnapshotList,
    testImageList,
  ];

  for (const test of tests) {
    try {
      await test();
    } catch (err) {
      console.log(`  ${RED}✗${NC} ${test.name}: ${err.message}`);
      failed++;
    }
  }

  console.log('\n=====================================');
  console.log('Test Summary');
  console.log('=====================================');
  console.log(`Passed: ${GREEN}${passed}${NC}`);
  console.log(`Failed: ${RED}${failed}${NC}`);
  console.log('=====================================');

  process.exit(failed > 0 ? 1 : 0);
}

main();

File: clients/java/sync/tests/TestFunctional.java
/**
 * This is free software for the public good of a permacomputer hosted at
 * permacomputer.com, an always-on computer by the people, for the people.
 * One which is durable, easy to repair, & distributed like tap water
 * for machine learning intelligence.
 *
 * The permacomputer is community-owned infrastructure optimized around
 * four values:
 *
 *   TRUTH      First principles, math & science, open source code freely distributed
 *   FREEDOM    Voluntary partnerships, freedom from tyranny & corporate control
 *   HARMONY    Minimal waste, self-renewing systems with diverse thriving connections
 *   LOVE       Be yourself without hurting others, cooperation through natural law
 *
 * This software contributes to that vision by enabling code execution across 42+ programming languages through a unified interface, accessible to all.
 * Code is seeds to sprout on any abandoned technology.
 *
 * UN Java SDK - Functional Tests
 *
 * Tests library functions against real API.
 * Requires: UNSANDBOX_PUBLIC_KEY, UNSANDBOX_SECRET_KEY
 *
 * Usage:
 *   javac -cp src tests/TestFunctional.java && java -cp src:tests TestFunctional
 */

import java.io.IOException;
import java.util.*;

public class TestFunctional {

    private static int passed = 0;
    private static int failed = 0;

    private static final String GREEN = "\033[32m";
    private static final String RED = "\033[31m";
    private static final String NC = "\033[0m";

    private static void check(boolean condition, String msg) {
        if (condition) {
            System.out.println("  " + GREEN + "✓" + NC + " " + msg);
            passed++;
        } else {
            System.out.println("  " + RED + "✗" + NC + " " + msg);
            failed++;
        }
    }

    private static void testHealthCheck() {
        System.out.println("\nTesting healthCheck()...");
        boolean result = Un.healthCheck();
        check(true, "healthCheck completed without exception");
    }

    private static void testValidateKeys() throws IOException {
        System.out.println("\nTesting validateKeys()...");
        Map<String, Object> info = Un.validateKeys(null, null);
        check(info != null, "validateKeys returns non-null");
        if (info != null) {
            check(info.containsKey("valid"), "result has 'valid' key");
            check(Boolean.TRUE.equals(info.get("valid")), "keys are valid");
            Object tier = info.get("tier");
            if (tier != null) System.out.println("    tier: " + tier);
        }
    }

    private static void testGetLanguages() throws IOException {
        System.out.println("\nTesting getLanguages()...");
        List<String> langs = Un.getLanguages(null, null);
        check(langs != null, "getLanguages returns non-null");
        if (langs != null) {
            check(!langs.isEmpty(), "at least one language returned");
            check(langs.contains("python"), "python is in languages list");
            System.out.println("    Found " + langs.size() + " languages");
        }
    }

    private static void testExecute() throws IOException {
        System.out.println("\nTesting executeCode()...");
        Map<String, Object> result = Un.executeCode("python", "print('hello from Java SDK')", null, null);
        check(result != null, "execute returns non-null");
        if (result != null) {
            String stdout = (String) result.get("stdout");
            check(stdout != null && stdout.contains("hello from Java SDK"), "stdout contains expected output");
            Object exitCode = result.get("exit_code");
            check(exitCode != null && ((Number) exitCode).intValue() == 0, "exit code is 0");
        }
    }

    private static void testExecuteError() throws IOException {
        System.out.println("\nTesting executeCode() with error...");
        Map<String, Object> result = Un.executeCode("python", "import sys; sys.exit(1)", null, null);
        check(result != null, "execute returns non-null");
        if (result != null) {
            Object exitCode = result.get("exit_code");
            check(exitCode != null && ((Number) exitCode).intValue() == 1, "exit code is 1");
        }
    }

    private static void testSessionList() throws IOException {
        System.out.println("\nTesting listSessions()...");
        List<Map<String, Object>> sessions = Un.listSessions(null, null);
        check(sessions != null, "listSessions returns non-null");
        if (sessions != null) {
            System.out.println("    Found " + sessions.size() + " sessions");
        }
    }

    private static void testSessionLifecycle() throws IOException {
        System.out.println("\nTesting session lifecycle (create, destroy)...");
        Map<String, Object> session = Un.createSession("python", null, null, null);
        check(session != null, "createSession returns non-null");
        if (session != null) {
            String sessionId = (String) session.get("id");
            check(sessionId != null, "session has id");
            System.out.println("    session_id: " + sessionId);

            if (sessionId != null) {
                Un.deleteSession(sessionId, null, null);
                check(true, "deleteSession completed");
            }
        }
    }

    private static void testServiceList() throws IOException {
        System.out.println("\nTesting listServices()...");
        List<Map<String, Object>> services = Un.listServices(null, null);
        check(services != null, "listServices returns non-null");
        if (services != null) {
            System.out.println("    Found " + services.size() + " services");
        }
    }

    private static void testSnapshotList() throws IOException {
        System.out.println("\nTesting listSnapshots()...");
        List<Map<String, Object>> snapshots = Un.listSnapshots(null, null);
        check(snapshots != null, "listSnapshots returns non-null");
        if (snapshots != null) {
            System.out.println("    Found " + snapshots.size() + " snapshots");
        }
    }

    private static void testImageList() throws IOException {
        System.out.println("\nTesting listImages()...");
        List<Map<String, Object>> images = Un.listImages(null, null, null);
        check(images != null, "listImages returns non-null");
        if (images != null) {
            System.out.println("    Found " + images.size() + " images");
        }
    }

    public static void main(String[] args) {
        System.out.println("=====================================");
        System.out.println("UN Java SDK - Functional Tests");
        System.out.println("Testing against real API");
        System.out.println("=====================================");

        String pk = System.getenv("UNSANDBOX_PUBLIC_KEY");
        String sk = System.getenv("UNSANDBOX_SECRET_KEY");

        if (pk == null || sk == null || pk.isEmpty() || sk.isEmpty()) {
            System.out.println("\n\033[33mSKIP: UNSANDBOX_PUBLIC_KEY and UNSANDBOX_SECRET_KEY not set\033[0m");
            System.exit(0);
        }

        try { testHealthCheck(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testValidateKeys(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testGetLanguages(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testExecute(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testExecuteError(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testSessionList(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testSessionLifecycle(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testServiceList(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testSnapshotList(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }
        try { testImageList(); } catch (Exception e) { System.out.println("  " + RED + "✗ " + e.getMessage() + NC); failed++; }

        System.out.println("\n=====================================");
        System.out.println("Test Summary");
        System.out.println("=====================================");
        System.out.println("Passed: " + GREEN + passed + NC);
        System.out.println("Failed: " + RED + failed + NC);
        System.out.println("=====================================");

        System.exit(failed > 0 ? 1 : 0);
    }
}
